### PR TITLE
Fix: #1655 - Added working Device Check API that is compliant with the DeviceCheck Server

### PR DESCRIPTION
**Pull request by Brandon-T**
- Security Review: https://github.com/brave/security/issues/13
- Spec: https://github.com/brave-intl/bat-captcha/pull/45
- Added working Device Check API that is compliant with the DeviceCheck Server
- Added PEM formatting (ASN.1 DER - ANSI X9.62) for Elliptic Curve Keys following the RFC spec.
- Removed biometric requirements from the Keys.
- Added signing and verifying of challenge nonce with the key as per Brave spec.
- Added example flow for Device Check from Brave spec.
- Made URLs compliant with Brave server instead of RFC spec.

**Hello World:**
```swift
//Sample Flow:
//generate Token -> generateEnrollment -> registerDevice
//getAttestation -> setAttestation (verify it)
// Verify if we failed or not.. and do something with the result..

class DeviceCheckExample {
  private let client = DeviceCheckClient()
  private var token = ""
  private let paymentId = UUID().uuidString //This comes from Brave Rewards..
  private var completion: ((_ error: Error?) -> Void)?
  
  // Starts the device check flow with the server
  func startHandshake(_ completion: @escaping (Error?) -> Void) {
    self.completion = completion
    self.generateToken()
  }

  deinit { self.completion = nil }
  
  // Generates a device check token using the iOS Apple APIs.
  // I do not store this token anywhere atm.
  private func generateToken() {
    client.generateToken { [weak self] token, error in
      guard let self = self else { return }
      
      if let error = error {
        self.completion?(error)
        return
      }
      
      self.token = token
      self.generateEnrollment()
    }
  }
  
  // Generates an enrollment structure using the token and making a call to the server.
  private func generateEnrollment() {
    if token.isEmpty {
      self.completion?("Invalid Device Check Token")
      return
    }
    
    client.generateEnrollment(paymentId: paymentId, token: token) { [weak self] registration, error in
      guard let self = self else { return }
      
      if let error = error {
        self.completion?(error)
        return
      }
      
      self.registerDevice(registration)
    }
  }
  
  // Registers the device with the server
  private func registerDevice(_ registration: DeviceCheckRegistration?) {
    guard let registration = registration else {
      self.completion?("Error registering device - Unknown Reason")
      return
    }
    
    client.registerDevice(enrollment: registration) { [weak self] error in
      guard let self = self else { return }
      
      if let error = error {
        self.completion?(error)
        return
      }
      
      self.getAttestation()
    }
  }
  
  // Retrieves a challenge-nonce from the server
  private func getAttestation() {
    client.getAttestation(paymentId: paymentId) { [weak self] attestation, error in
      guard let self = self else { return }
      
      if let error = error {
        self.completion?(error)
        return
      }
      
      self.setAttestation(attestation)
    }
  }
  
  // Responds to the challenge and sends it to the server.
  private func setAttestation(_ attestation: AttestationBlob?) {
    guard let attestation = attestation else {
      self.completion?("Error retrieving attestation - Unknown Reason")
      return
    }
    
    client.setAttestation(nonce: attestation.nonce) { [weak self] error in
      guard let self = self else { return }
      
      if let error = error {
        self.completion?(error)
        return
      }
      
      self.completion?(nil)
    }
  }
}
```

**Hello World Usage:**
```swift
DeviceCheckExample().startHandshake { error in
  if let error = error {
    log.error(error) //Failed to retrieve attestation for this device.. Do something..
    return
  }
  
  //Attestation has been successful so do something else now.. claim the grant I guess?..
}
```

## Summary of Changes

This pull request fixes issue https://github.com/brave/brave-ios/issues/1655

## Submitter Checklist:

- [ ] *Unit Tests* are updated to cover new or changed functionality
- [x] User-facing strings use `NSLocalizableString()`

## Reviewer Checklist:

- [ ] Issues include necessary QA labels:
  - `QA/(Yes|No)`
  - `release-notes/(include|exclude)`
  - `bug` / `enhancement`
- [ ] Necessary [security reviews](https://github.com/brave/security/issues/new/choose) have taken place.
- [ ] Adequate unit test coverage exists to prevent regressions.
- [ ] Adequate test plan exists for QA to validate (if applicable).
- [ ] Issue is assigned to a milestone (should happen at merge time).

### CODE REVIEW

File: Client/DeviceCheck/Cryptography.swift
Comment (iccub) [inline Client/DeviceCheck/Cryptography.swift]: Do you think it's better to crash the app here instead of unwrapping this and return false?

File: Client/DeviceCheck/Cryptography.swift
Comment (jumde) [inline Client/DeviceCheck/Cryptography.swift]: Do you expect the private key to be used outside this file? If not we can make it private and only expose the utility functions like `sign` and `getPublicKey`.

Comment (Brandon-T) [reply Client/DeviceCheck/Cryptography.swift]: True. I will fix it. The private key is not needed outside this file but also, the private key can't be exported as well so it doesn't hurt :)

But yeah, good point. It will be fixed.

File: Client/DeviceCheck/DeviceCheck.swift
Comment (jumde) [inline Client/DeviceCheck/DeviceCheck.swift]: https

Comment (Brandon-T) [reply Client/DeviceCheck/DeviceCheck.swift]: I got the url from slack:

image

If I use `https`, I get an error atm. I'm assuming it's because it's dev which is why I never mentioned it and I'm thinking production and staging will be different than dev.

File: Client/DeviceCheck/Cryptography.swift
Comment (iccub) [inline Client/DeviceCheck/Cryptography.swift]: All `CryptographyError` return code = -1. I think we could hardcode it, or at least add as a default init parameter

File: Client/DeviceCheck/Cryptography.swift
Comment (iccub) [inline Client/DeviceCheck/Cryptography.swift]: Unwrap and throw error if can't be unwrapped

File: Client/DeviceCheck/Cryptography.swift
Comment (iccub) [inline Client/DeviceCheck/Cryptography.swift]: This query is the same as in `isKeyRequiringBiometrics` perhaps we could extract it?

File: Client/DeviceCheck/Cryptography.swift
Comment (iccub) [inline Client/DeviceCheck/Cryptography.swift]: Remove force cast

File: Client/DeviceCheck/Cryptography.swift
Comment (iccub) [inline Client/DeviceCheck/Cryptography.swift]: Remove force cast if it's better to fail safely

File: Client/DeviceCheck/Cryptography.swift
Comment (iccub) [inline Client/DeviceCheck/Cryptography.swift]: Remove force unwrapping

File: Client/DeviceCheck/Cryptography.swift
Comment (iccub) [inline Client/DeviceCheck/Cryptography.swift]: remove force cast

**Review by iccub — CHANGES_REQUESTED** [review]
Mostly some concerns about force casting and force unwrapping, I don't think we should use so many force casts in this code.
If there is a reason for it please let me know

File: Client/DeviceCheck/Cryptography.swift
Comment (Brandon-T) [reply Client/DeviceCheck/Cryptography.swift]: image

Screen Shot 2019-10-24 at 11 51 58 AM

This is a bug in Xcode and it can't be done any other way. There's no way to unwrap a `CFTypeRef?` as another `CF` type. You will always end up with a force somewhere :(

I've tried to `if let`, I have tried `guard let`, I have tried optional cast and ternary unwrap.. None of them work and will give the above errors. That's why I have to force cast it to an optional type and then unwrap that optional type :(

Comment (iccub) [reply Client/DeviceCheck/Cryptography.swift]: Lol XCode really doesn't like you :D
Would be good to put a comment about it somewhere in code

Comment (Brandon-T) [reply Client/DeviceCheck/Cryptography.swift]: I fixed it by first checking what type was returned via its TypeId and then force-casting. This should provide the safety we're looking for.

File: Client/DeviceCheck/DeviceCheck.swift
Comment (iccub) [inline Client/DeviceCheck/DeviceCheck.swift]: Remove `code`

File: Client/DeviceCheck/DeviceCheck.swift
Comment (iccub) [inline Client/DeviceCheck/DeviceCheck.swift]: Remove `code`

File: Client/DeviceCheck/DeviceCheck.swift
Comment (iccub) [inline Client/DeviceCheck/DeviceCheck.swift]: Remove `code`

File: Client/DeviceCheck/DeviceCheck.swift
Comment (iccub) [inline Client/DeviceCheck/DeviceCheck.swift]: Remove `code`

File: Client/DeviceCheck/DeviceCheck.swift
Comment (iccub) [inline Client/DeviceCheck/DeviceCheck.swift]: Remove `code`

File: Client/DeviceCheck/DeviceCheck.swift
Comment (iccub) [inline Client/DeviceCheck/DeviceCheck.swift]: Is this code still needed?

File: Client/DeviceCheck/DeviceCheck.swift
Comment (iccub) [inline Client/DeviceCheck/DeviceCheck.swift]: Still not sure or can this comment be updated?

File: Client/DeviceCheck/Cryptography.swift
Comment (iccub) [inline Client/DeviceCheck/Cryptography.swift]: remove if not needed

**Review by iccub — CHANGES_REQUESTED** [review]
Looks good in general, need some cleanup - removing commented out code

**Review by iccub — APPROVED** [review]
Goood, do we wait with this PR until server side is complete?

**Comment by Brandon-T** [comment]
> Goood, do we wait with this PR until server side is complete?

The server side works with this PR right now. We're currently waiting on grants to be completed but I'm working on integrating this into the Wallet creation atm (in a local branch) & waiting on Security team to give the ok for this PR. Once it's merged, we can completely integrate it into wallet and make a PR for that. Then we wait again for the grant stuff to be complete & continue integration.

**Review by jseidl — APPROVED** [review]
Reviewed the code for common general programming issues but @riastradh-brave is a better suited person for attesting the cryptography-related code.

All looks good for me. Good job folks!

**Review by jumde — APPROVED** [review]
lgtm

**Pull request merged**
